src/WebJobs.Script/Description/CSharp/CSharpFunctionInvoker.cs

```diff
@@ -0,0 +1,422 @@
+            {
+                _reloadScript();
+            }
+            else if (string.Compare("project.json", e.Name, true) == 0)
+            {
+                _restorePackages();
+            }
+        }
+
+        private void ReloadScript()
+        {
+            // Reset cached function
+            _function = null;
+            _fileTraceWriter.Verbose(string.Format("Script for function '{0}' changed. Reloading.", _functionMetadata.Name));
+
+            _fileTraceWriter.Verbose("Compiling function script.");
+            var stopwatch = new Stopwatch();
+            stopwatch.Start();
+
+            Script<object> script = CreateScript();
+            ImmutableArray<Diagnostic> compilationResult = script.Compile();
+
+            stopwatch.Stop();
+            _fileTraceWriter.Verbose($"Compilation completed ({stopwatch.ElapsedMilliseconds} ms) .");
+
+            foreach (var diagnostic in compilationResult)
+            {
+                var traceEvent = new TraceEvent(GetTraceLevelFromDiagnostic(diagnostic), diagnostic.ToString());
+                _fileTraceWriter.Trace(traceEvent);
+            }
+        }
+
+        private void RestorePackages()
+        {
+            _fileTraceWriter.Verbose("Restoring script packages.");
+
+            _metadataResolver.RestorePackagesAsync()
+                .ContinueWith(t =>
+                {
+                    if (t.IsFaulted)
+                    {
+                        _fileTraceWriter.Verbose("Package restore failed:");
+                        _fileTraceWriter.Verbose(t.Exception.ToString());
+                        return;
+                    }
+
+                    _fileTraceWriter.Verbose("Packages restored.");    
+                    _reloadScript();
+                });
+        }
+
+        public override async Task Invoke(object[] parameters)
+        {
+            object input = parameters[0];
+            TraceWriter traceWriter = (TraceWriter)parameters[1];
+            IBinder binder = (IBinder)parameters[2];
+            ExecutionContext functionExecutionContext = (ExecutionContext)parameters[3];
+
+            try
+            {
+                _fileTraceWriter.Verbose("Function started");
+
+                // if there are any binding parameters in the output bindings,
+                // parse the input as json to get the binding data
+                var bindingData = new Dictionary<string, string>();
+                if (_outputBindings.Any(p => p.HasBindingParameters) ||
+                    _inputBindings.Any(p => p.HasBindingParameters))
+                {
+                    bindingData = GetBindingData(input);
+                }
+                bindingData["InvocationId"] = functionExecutionContext.InvocationId.ToString();
+
+                MethodInfo function = GetFunctionTarget();
+
+                object[] arguments = await BindArgumentsAsync(binder, function, bindingData, parameters);
```

Comment:
rename to ProcessInputBindings? So the pattern is consistent with other invokers? (unless you have a reason why that name doesn't make sense)